connectors/sources/box.py

```diff
@@ -0,0 +1,445 @@
+                    )
+        except Exception as exception:
+            raise TokenError(
+                f"Error while generating access token. Please verify that provided configurations are correct. Exception {exception}."
+            ) from exception
+
+
+class BoxClient:
+    def __init__(self, configuration):
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self._logger = logger
+        self.is_enterprise = configuration["is_enterprise"]
+        self._http_session = aiohttp.ClientSession(
+            base_url="https://api.box.com", raise_for_status=True
+        )
+        self.token = AccessToken(
+            configuration=configuration, http_session=self._http_session
+        )
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    async def _put_to_sleep(self, exception):
+        retry_after = int(exception.headers.get("retry-after", 5))
+        self._logger.debug(
+            f"Connector will attempt to retry after {retry_after} seconds."
+        )
+        await self._sleeps.sleep(retry_after)
+        raise Exception("Rate limit exceeded.")
+
+    async def _handle_client_errors(self, exception):
+        match exception.status:
+            case 401:
+                await self.token._set_access_token()
+                raise
+            case 429:
+                await self._put_to_sleep(exception=exception)
+            case 404:
+                raise NotFound(f"Resource Not Found. Error: {exception}")
+            case _:
+                raise
+
+    @retryable(
+        retries=RETRIES,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+        skipped_exceptions=NotFound,
+    )
+    async def get(self, url, headers, params=None):
+        try:
+            access_token = await self.token.get()
+            headers.update({"Authorization": f"Bearer {access_token}"})
+            return await self._http_session.get(url=url, headers=headers, params=params)
+        except ClientResponseError as exception:
+            await self._handle_client_errors(exception=exception)
+        except Exception:
+            raise
+
+    async def paginated_call(self, url, params, headers):
+        try:
+            offset = 0
+            while True:
+                params.update({"offset": offset, "limit": FETCH_LIMIT})
+                response = await self.get(url=url, headers=headers, params=params)
+                json_response = await response.json()
+                total_count = json_response.get("total_count")
+                for doc in json_response.get("entries"):
+                    yield doc
+                if offset >= total_count:
+                    break
+                offset += FETCH_LIMIT
+        except Exception:
+            raise
+
+    async def ping(self):
+        await self.get(url=ENDPOINTS["PING"], headers={})
+
+    async def close(self):
```

Comment:
Should we also clean up the queue and the fetchers?

Reply:
In the consumer method, we are getting all the documents from the queue. So, it will empty at the end, and fetchers are nothing but tasks, so i think there is no need to clean it.

Reply:
When the framework cancels the sync job it should clean up all the resources (HTTP connections, queues, etc.). The framework will be waiting for the job to be canceled only for a specific time (5 seconds by default) so cleaning the queue might be useful to avoid "kill" stop and do everything gracefully.

Reply:
I have cleaned the queue at the `close` method of `BoxDataSource` as we are using queue in `BoxDataSource` [here](https://github.com/elastic/connectors-python/pull/1631/commits/e5e9edd9a3956e13ba02709106a2b601c1f9ee11). Let me know if I missed something.